### ISSUE TITLE
wrong error message with if-let return type mismatch

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#35685]
This function

``` rust
fn foo() -> bool { if let Some(()) = Some(()) { } else { } }
```

Produces E0308 (correct) with the message "`if-let` arms have incompatible types" (nonsense).

**Comment by durka** [comment]
cc @jonathandturner @GuillaumeGomez

**Comment by nagisa** [comment]
This is what is reported for `match` version of this:

```
error: match arms have incompatible types [--explain E0308]
 --> <anon>:2:5
2 |>     match Some(()) {
  |>     ^ expected bool, found ()
note: expected type `_`
note:    found type `()`
note: match arm with an incompatible type
 --> <anon>:3:21
3 |>         Some(()) => { }
  |>                     ^^^
```

I feel like this falls in the same category, and the issue could be fixed by highlighting the branch with conflicting type.

**Comment by durka** [comment]
Makes sense, since `if let` is desugared to `match` pretty early on. But IMO that error message is wrong too. The arms are compatible with each other, but they don't match the return type of the function. Also, it says "expected type `_`" and I thought we got rid of that.

**Comment by nagisa** [comment]
I think there’s multiple ways to interpret the message. I guess you’re interpreting it as

> match/if-let arms have incompatible types (between branches, i.e. one branch has mismatching type with another)

whereas another entirely valid interpretation is

> match/if-let arm(s) have incompatible types (compared to what they ought to be given inferred type of the whole expression)

**Comment by durka** [comment]
Yes, you're right. Let's add those parentheticals as NOTEs in the error
message :)

**Comment by Mark-Simulacrum** [comment]
Technically, today's message is correct. Ideally we'd error on the whole expression instead of each arm, but I'm going to close since we no longer say anything about incompatible types of arms.

```
error[E0308]: mismatched types
 --> test.rs:1:47
  |
1 | fn foo() -> bool { if let Some(()) = Some(()) { } else { } }
  |                                               ^^^ expected bool, found ()
  |
  = note: expected type `bool`
             found type `()`

error[E0308]: mismatched types
 --> test.rs:1:56
  |
1 | fn foo() -> bool { if let Some(()) = Some(()) { } else { } }
  |                                                        ^^^ expected bool, found ()
  |
  = note: expected type `bool`
             found type `()`

error: aborting due to 2 previous errors
```
